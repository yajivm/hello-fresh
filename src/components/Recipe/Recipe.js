import React, { useState, useEffect } from "react";
import Card from '../../components/Card';
import { recipes } from "../../mockData/data";
import { useNavigateParams } from "../../utils/helper";
import { DEFAULT_PAGE_NUMBER, ITEMS_COUNT_PER_PAGE } from '../../utils/constant';
import { BackIc, ForwardIc } from "../Icons";
import { GridWrapper, GridContainer, Grid, GreyText, Container, PaginatioNumber, PaginationWrapper, Icon } from './Recipe.styles';

const Recipe = () => {
  const navigate = useNavigateParams();
  const [recipeList, setRecipeList] = useState([]);
  const [pageNumber, setPageNumber] = useState(DEFAULT_PAGE_NUMBER);
  const [showPagination, setShowPagination] = useState(false);
  const [showBackIc, setShowBackIc] = useState(false);
  const [showForwardIc, setShowForwardIc] = useState(!(recipeList.length > ITEMS_COUNT_PER_PAGE));

  const onClickCard = id => {
    navigate(`/recipe-details`, `id=${id}`);
  }

  const deleteCard = id => {
    const filteredData = recipeList.filter(item => item.id !== id);
    setRecipeList(filteredData);
  }

  const editCard = id => {
    console.log('====>', id)
  }

  const filteredItemsForPage = () => {
    let recipesData = [];
    let paginationStaus = false;
    if (recipes.length > ITEMS_COUNT_PER_PAGE) {
      const filteredData = recipes.filter((item, ind) => ind < ITEMS_COUNT_PER_PAGE - 1);
      recipesData = filteredData;
      paginationStaus = true;
    } else {
      recipesData = recipes;
    }
    console.log('=====>', recipesData.length)
    setRecipeList(recipesData);
    setShowPagination(paginationStaus);
  }

  useEffect(() => {
    filteredItemsForPage();
  }, []);

  const numPages = () => {
    return Math.ceil(recipes.length / ITEMS_COUNT_PER_PAGE);
  }

  const onChangePage = page => {
    let recipesData = [];
    if (page < 1) {
      page = 1;
    }
    if (page > numPages()) page = numPages();

    for (var i = (page-1) * ITEMS_COUNT_PER_PAGE; i < (page * ITEMS_COUNT_PER_PAGE) && i < recipes.length; i++) {
      recipesData.push(recipes[i]);
    }
    
    if (page === 1) {
      setShowBackIc(false)
    } else setShowBackIc(true)

    console.log(page === numPages())
    if (page === numPages()) {
      setShowForwardIc(false)
    } else setShowForwardIc(true)

    setRecipeList(recipesData);
    setPageNumber(page);
  }

  const onClickPagination = type => {
    let number = pageNumber;
    switch (type) {
      case 'Back':
        number--;
        break;
      case 'Forward':
        number++;
        break;
      default:
        break;
    }
    onChangePage(number);
  }

  return (
    <Container>
      <GridWrapper>
        <GridContainer>
          {
            (recipeList && recipeList.length > 0) ?
              recipeList.map(({ title, description, img, id }, ind) => (
                <Grid key={`receipie-${ind}`} count={4}>
                  <Card title={title} description={description} imageSrc={img} id={id} onClickCard={onClickCard} showActionIcons onClickDeleteIcon={deleteCard} onClickEditIcon={editCard} />
                </Grid>
              ))
            :
              <GreyText>No Recipe Available</GreyText>
          }
        </GridContainer>
      </GridWrapper>
      {showPagination && (
        <PaginationWrapper>
          <Icon onClick={() => onClickPagination('Back')} isIcHidden={showBackIc}>
            <BackIc />
          </Icon>
          <PaginatioNumber>{pageNumber}</PaginatioNumber>
          <Icon onClick={() => onClickPagination('Forward')} isIcHidden={showForwardIc}>
            <ForwardIc />
          </Icon>
        </PaginationWrapper>
      )}
    </Container>
  )
}

export default Recipe;