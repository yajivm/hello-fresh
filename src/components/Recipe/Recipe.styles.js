import styled from 'styled-components';

export const GridWrapper = styled.div`
  width: 100%;
  overflow: hidden;
`

export const GridContainer = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
`

export const Grid = styled.div`
  width: ${({ count }) => count === 4 ? `23.2%` : `31.55%`};
  margin: 0 10px 30px;
  cursor: pointer;

  ${({ count }) => count !== 4 ?
    `
      @media(max-width: 1145px) {
        width: 48%;
      }
      @media(max-width: 1000px) {
        width: 47.5%;
      }
      @media(max-width: 800px) {
        max-width: 100%;
        margin: 0 auto 10px;
        width: 90vw;
      }    
    `
  :
    `@media(max-width: 1110px) {
        width: 47.5%;
      }

      @media(max-width: 800px) {
        max-width: 100%;
        margin: 0 auto 10px;
        width: 90vw;
      }    
    `
  }
`

export const GreyText = styled.p`
  font-size: 40px;
  line-height: 52px;
  text-align: center;
  width: 100%;
  color: rgba(0, 0, 0, 0.2);
  margin-top: 40vh;
`

export const Container = styled.div``