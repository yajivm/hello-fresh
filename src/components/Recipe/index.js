import Recipie from './Recipe';

export default Recipie;