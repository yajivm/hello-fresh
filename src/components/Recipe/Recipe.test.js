import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react';
import Recipe from '.'

const mockUseLocationValue = {
    pathname: "/testroute",
    search: '',
    hash: '',
    state: null
}

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => {
  const originalModule = jest.requireActual('react-router-dom');
  return {
    ...originalModule,
    useLocation: jest.fn().mockImplementation(() => {
        return mockUseLocationValue;
    }),
    useNavigate: () => mockedUsedNavigate,
  };
});

describe('Recipe', () => {

  it('should render component', () => {
    const { container } = render(<Recipe />)

    expect(container).toMatchSnapshot();
  })

  it('should remove the card when click the delete icon over the card', () => {
    render(<Recipe />)

    expect(screen.getByTestId('card-content-0')).toBeInTheDocument()
    expect(screen.getByTestId('delete-ic-0')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('delete-ic-0'))

    expect(screen.queryByTestId('card-content-0')).not.toBeInTheDocument()
  })
})