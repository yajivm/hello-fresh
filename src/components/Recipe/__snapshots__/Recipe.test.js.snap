// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe should render component 1`] = `
<div>
  <div
    class="sc-breuTD"
  >
    <div
      class="sc-kDDrLX fvofKy"
    >
      <div
        class="sc-iqcoie kyVYqQ"
      >
        <div
          class="sc-crXcEl gGybZp"
        >
          <div
            class="sc-gKXOVf hdSzut"
            data-testid="0"
          >
            <div
              class="sc-eCYdqJ bjhUVy"
            >
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="delete-ic-0"
              >
                <svg
                  fill="currentColor"
                  height="1.5em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </div>
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="edit-ic-0"
              >
                <svg
                  fill="currentColor"
                  height="1.2em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1.2em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="sc-iBkjds"
              data-testid="card-content-0"
            >
              <img
                alt="card-image-0"
                class="sc-bczRLJ blDScb"
                data-testid="card-image-0"
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/pesto-parmesan-crusted-chicken-5f50e32e.jpg"
              />
              <div
                class="sc-gsnTZi hhwjBW"
              >
                <h3
                  class="sc-dkzDqf knkAIX"
                  data-testid="card-title-0"
                >
                  Easy Pesto-Crusted Chicken
                </h3>
                <div
                  class="sc-hKMtZM"
                  data-testid="card-description-0"
                >
                  with Roast Sweet Potato & Rainbow Salad
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-crXcEl gGybZp"
        >
          <div
            class="sc-gKXOVf hdSzut"
            data-testid="1"
          >
            <div
              class="sc-eCYdqJ bjhUVy"
            >
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="delete-ic-1"
              >
                <svg
                  fill="currentColor"
                  height="1.5em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </div>
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="edit-ic-1"
              >
                <svg
                  fill="currentColor"
                  height="1.2em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1.2em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="sc-iBkjds"
              data-testid="card-content-1"
            >
              <img
                alt="card-image-1"
                class="sc-bczRLJ blDScb"
                data-testid="card-image-1"
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/japanese-style-beef-sesame-veggies-cb7e5629.jpg"
              />
              <div
                class="sc-gsnTZi hhwjBW"
              >
                <h3
                  class="sc-dkzDqf knkAIX"
                  data-testid="card-title-1"
                >
                  Japanese-Style Beef & Veggies
                </h3>
                <div
                  class="sc-hKMtZM"
                  data-testid="card-description-1"
                >
                  with Rice & Coconut Sweet Chilli Mayo
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-crXcEl gGybZp"
        >
          <div
            class="sc-gKXOVf hdSzut"
            data-testid="2"
          >
            <div
              class="sc-eCYdqJ bjhUVy"
            >
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="delete-ic-2"
              >
                <svg
                  fill="currentColor"
                  height="1.5em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </div>
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="edit-ic-2"
              >
                <svg
                  fill="currentColor"
                  height="1.2em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1.2em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="sc-iBkjds"
              data-testid="card-content-2"
            >
              <img
                alt="card-image-2"
                class="sc-bczRLJ blDScb"
                data-testid="card-image-2"
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/asian-pork-meatball-tacos-fc8c6b21.jpg"
              />
              <div
                class="sc-gsnTZi hhwjBW"
              >
                <h3
                  class="sc-dkzDqf knkAIX"
                  data-testid="card-title-2"
                >
                  Asian Pork Meatball Tacos
                </h3>
                <div
                  class="sc-hKMtZM"
                  data-testid="card-description-2"
                >
                  with Sweet Chilli Glaze & Celery Slaw
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-crXcEl gGybZp"
        >
          <div
            class="sc-gKXOVf hdSzut"
            data-testid="3"
          >
            <div
              class="sc-eCYdqJ bjhUVy"
            >
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="delete-ic-3"
              >
                <svg
                  fill="currentColor"
                  height="1.5em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </div>
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="edit-ic-3"
              >
                <svg
                  fill="currentColor"
                  height="1.2em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1.2em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="sc-iBkjds"
              data-testid="card-content-3"
            >
              <img
                alt="card-image-3"
                class="sc-bczRLJ blDScb"
                data-testid="card-image-3"
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/harissa-honey-chicken-a7d73d96.jpg"
              />
              <div
                class="sc-gsnTZi hhwjBW"
              >
                <h3
                  class="sc-dkzDqf knkAIX"
                  data-testid="card-title-3"
                >
                  Chermoula-Honey Chicken
                </h3>
                <div
                  class="sc-hKMtZM"
                  data-testid="card-description-3"
                >
                  with Zesty Salsa, Garlic Rice & Almonds
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-crXcEl gGybZp"
        >
          <div
            class="sc-gKXOVf hdSzut"
            data-testid="4"
          >
            <div
              class="sc-eCYdqJ bjhUVy"
            >
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="delete-ic-4"
              >
                <svg
                  fill="currentColor"
                  height="1.5em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1.5em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </div>
              <div
                class="sc-jSMfEi ebJwAT"
                data-testid="edit-ic-4"
              >
                <svg
                  fill="currentColor"
                  height="1.2em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1.2em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="sc-iBkjds"
              data-testid="card-content-4"
            >
              <img
                alt="card-image-4"
                class="sc-bczRLJ blDScb"
                data-testid="card-image-4"
                src="https://img.hellofresh.com/c_fill,f_auto,fl_lossy,h_432,q_auto,w_768/hellofresh_s3/image/chorizo-potato-tray-bake-f377190c.jpg"
              />
              <div
                class="sc-gsnTZi hhwjBW"
              >
                <h3
                  class="sc-dkzDqf knkAIX"
                  data-testid="card-title-4"
                >
                  Chorizo & Veggie Traybake
                </h3>
                <div
                  class="sc-hKMtZM"
                  data-testid="card-description-4"
                >
                  with Spinach, Fetta & Almonds
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-ftvSup bWaAUx"
      data-testid="pagination-wrapper"
    >
      <div
        class="sc-papXJ ifiNAn"
        data-testid="back-ic"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.707 16.707a1 1 0 01-1.414 0l-6-6a1 1 0 010-1.414l6-6a1 1 0 011.414 1.414L5.414 9H17a1 1 0 110 2H5.414l4.293 4.293a1 1 0 010 1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <p
        class="sc-jqUVSM eugsYg"
        data-testid="page-number-title"
      >
        1
      </p>
      <div
        class="sc-papXJ gIUWur"
        data-testid="forward-ic"
      >
        <svg
          fill="currentColor"
          height="1.5em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10.293 3.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
